### DISCUSSION TITLE
Lower sd clock speed for fake blue pills

### DISCUSSION THREAD

**Discussion by marcelv-3** [erichelgeson/BlueSCSI-v1#58] [General]
I have some fake blue pills which will not work with the standard SPI_FULL_SPEED 50 MHz
Result is not recognising the SD card and the led blinking 5 times.

My fake versions max out at 35 MHz, I added the following code so the max speed is determined at startup.

Could someone check this with their fake blue pills and if working could this be an addition to the code ?

```c++
  // clock = 36MHz , about 4Mbytes/sec

  //lower maxspeed for fake blue pills
  int i = 50;
  bool sdReady = false;
  while (i >= 32 && !sdReady)
  {
    if(SD.begin(SdSpiConfig(PA4, DEDICATED_SPI, SD_SCK_MHZ(i), &SPI))) {
      sdReady = true;
    }
    else
    {
      i--;
    }
  }

  if(!sdReady) {
#if DEBUG > 0
    Serial.println("SD initialization failed!");
#endif
    noSDCardFound();
  }
  initFileLog();

  LOG_FILE.print("SD1_Config ");
  LOG_FILE.print(i);
  LOG_FILE.println(" Mhz");
  LOG_FILE.sync();

  readSCSIDeviceConfig();
  readSDCardInfo();
```

**Comment by erichelgeson** [comment]
Hey, great idea! Could you provide some more details and a photo of the chip this helped you with along with the exact SD card you're using? So far the only issue we've seen is with [WCH using SanDisk SD cards](https://github.com/erichelgeson/BlueSCSI/wiki/SD-Cards#sandisk-and-wch-stm32-clones).

**Comment by marcelv-3** [comment]
The problem is in the Fake STM32 blue pill, not the SD card, Have tried multiple SD cards.

results for the fake STM32 from Blinky Test:

** Blinky Test **
** Keir Fraser <keir.xen@gmail.com>
** https://github.com/keirf/Greaseweazle
Serial = 57c0:0498:3309:1100:07e4:1506
Flash Size = 64kB (STM32F103x8 Medium-Density)
Device ID = 0x0410
Revision  = 0x2003
**WARNING**: Device returned valid IDCODE! Fake?
Testing I2C1... [Fake Chip?] **FAILED**
Testing I2C2... [Fake Chip?] **FAILED**
Testing SPI1... OK
Testing SPI2... OK
Testing TIM1... OK
Testing TIM2... OK
Testing TIM3... OK
Testing TIM4... OK
DMA Test #1... [Spurious IRQ: Fake Chip?] **FAILED**
DMA Test #2... [Spurious IRQ: Fake Chip?] [Bad Data] **FAILED**
DMA Test #3... [Spurious IRQ: Fake Chip?] [Bad Data] **FAILED**
DMA Test #4... [Spurious IRQ: Fake Chip?] [Bad Data] **FAILED**
Testing 64kB Flash... OK
Enable TIM4 IRQ... OK
Testing 20kB SRAM (endless loop)..
![FakeBluePill](https://user-images.githubusercontent.com/96822999/147685761-03b7ab63-e337-4bc4-a8ab-cd96e2825888.jpg)

**Reply by erichelgeson** [reply]
I have similarly marked chips (though markings are not a guarantee) and have had success at 50mhz. I'd take a look at the troubleshooting guide in our wiki (specifically about power). I still think that the patch you provided is a good one in case a bluepill or sd card has issues at higher speeds.

**Comment by marcelv-3** [comment]
The marking are not a quarantee, they may be similar but the one in the picture, you can see the markings are not centered. Original ones have markings with even spacing at the edges,


little side question for the original code:

#define SD1_CONFIG SdSpiConfig(PA4, DEDICATED_SPI, **SD_SCK_MHZ(SPI_FULL_SPEED)**, &SPI)
#define SPI_FULL_SPEED SD_SCK_MHZ(50)

so would SD_SCK_MHZ(SPI_FULL_SPEED) be the same as SD_SCK_MHZ(SD_SCK_MHZ(50)) which is different from SD_SCK_MHZ(50)  ??

**Reply by erichelgeson** [reply]
Yes looks like it is.
